beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -1,84 +1,303 @@
+
+Example usage:
+
+```typescript
+// Repository creation with rollback
+await ctx.checkpoint(
+  'repository-created',
+  async () => {
+    const { owner, repo } = ctx.input;
+    const repository = await githubClient.repos.create({ owner, repo });
+    return {
+      repositoryId: repository.id,
+      owner,
+      repo,
+    };
+  },
+  async data => {
+    // Rollback function using saved checkpoint data
+    const { owner, repo } = data;
+    await githubClient.repos.delete({ owner, repo });
+  },
+);
+```
+
+The checkpoint state will be extended to include a new status for rolled back operations. This could include new statuses for failed and rolled back operations.
+
+```typescript
+type CheckpointState =
+  | {
+      status: 'failed';
+      reason: string;
+    }
+  | {
+      status: 'success';
+      value: JsonValue;
+    }
+  | {
+      status: 'rolledback';
+      originalValue: JsonValue;
+    }
+  | {
+      status: 'rollback-failed';
+      originalValue: JsonValue;
+      reason: string;
+    };
+```
+
+This enhancement allows us to identify checkpoints that failed to rollback for potential manual intervention, thus tracking the complete lifecycle of a checkpoint (creation → potential rollback).
+
+### WorkflowRunner Enhancements
+
+The `NunjucksWorkflowRunner` will be modified to track successfully completed checkpoints with their rollback functions. When a scaffolder task fails, the system will:
+
+1. Query the database for completed checkpoints associated with the task
+2. Filter for checkpoints that have rollback functions defined
+3. Execute these rollback functions in reverse order
+
+The implementation will ensure that:
+
+- Rollbacks are executed in proper LIFO order across all actions
+- Each rollback operation is properly logged and tracked
+- **Failures in one rollback don't prevent other rollbacks from being attempted**
+
+```typescript
+// Pseudo-code for rollback execution
+for (const checkpoint of completedCheckpointsWithRollback.reverse()) {
+  try {
+    await checkpoint.rollback(checkpoint.data);
+    logger.info(`Successfully rolled back checkpoint: ${checkpoint.key}`);
+  } catch (error) {
+    logger.error(`Failed to rollback checkpoint ${checkpoint.key}: ${error}`);
+    // Continue with next rollback
+  }
+}
+```
+
+### Enabling Rollback for Checkpoints
+
+!!! warning Work in progress
+
+Rollbacks will be disabled by default and need to be explicitly enabled. We will add a flag to enable rollback functionality at the task level, with the option to override this at the action level.
+
+```yaml
+apiVersion: scaffolder.backstage.io/v1beta3
+kind: Template
+metadata:
+  name: example-template
+spec:
+  parameters:
+    # ...
+  steps:
+    - id: github-repo
+      name: Create GitHub Repository
+      action: github:repo:create
+      rollbackEnabled: true # Enable rollback for this action
```

Comment:
Is this just for testing for feature flagging? The `app-config.yaml` option seems fine right to opt into this behavior, do we really need to flag it per action?

Reply:
As a side note, should this be done automatically or should we have a button for rolling back in the UI?

Reply:
> The thought behind allowing an override for each action is perhaps one company does not allow deletion of repo under any circumstance. What if they wanted to disable that, but could leverage rollbacks else where?

Wondering if it should be `rollback: false` then an opt into disabling instead?